### ISSUE TITLE
Add optional support for Dkim signing

### ISSUE THREAD

**Issue by chaopeng** [bbottema/simple-java-mail#33]
extends Mailer to add sendDKIMMail()

**Comment by bbottema** [comment]
?

**Comment by chaopeng** [comment]
I saw a log: `Providing access to Session instance for emergency fall-back scenario. Please let us know why you need it.` I extends `org.codemonkey.simplejavamail.Mailer` and add a function to send DKIM mail. it need to convert the Message to DKIMMessage. 

```java
public final void sendDKIMMail(final Email email) throws MailException {
    if (validate(email)) {
        try {
            // fill and send wrapped mime message parts
            Message message = produceMimeMessage(email, getSession());
            convertToDKIMMessage(message, ...);
            message.saveChanges(); // some headers and id's will be set for this specific message
            Transport transport = getSession().getTransport();
            try {
                transport.connect();
                transport.sendMessage(message, message.getAllRecipients());
            } finally {
                transport.close();
            }
        } catch (final UnsupportedEncodingException e) {
            logger.error(e.getMessage(), e);
            throw new MailException(String.format("Encoding not accepted: %s", e.getMessage()), e);
        } catch (final MessagingException e) {
            logger.error(e.getMessage(), e);
            throw new MailException(String.format("Generic error: %s", e.getMessage()), e);
        }
    }
}
```

**Comment by bbottema** [comment]
Ahh, alright. Perhaps `Session session` should have been _protected_ instead of _private_, so that you can reference it directly in subclasses instead of through the bean getter method.

Btw, what is convertToDKIMMessage? Is that something this library might benefit from?

**Comment by chaopeng** [comment]
convertToDKIMMessage is a function to add DKIM signature to mail. I am using https://github.com/markenwerk/java-utils-mail-dkim to help me do this.

**Comment by bbottema** [comment]
Will you share your implementation? I'm not really familiar with DKIM.

**Comment by bbottema** [comment]
What about this, @chaopeng:

```
Email email = new Email();
email.sign(dkimPrivateKeyFileForSender, signingDomain, selector);
```

Which throws an exception if `DkimSigner` is missing on the classpath, making the maven dependency on Dkim optional.

**Comment by chaopeng** [comment]
Thank you @bbottema . maybe the method also support byte[]/String as key is better.

**Comment by bbottema** [comment]
@chaopeng, what would be the content of that? Would it be the key itself, or a filepath to the key?

**Comment by chaopeng** [comment]
the string/ byte[] will be the key.

**Comment by bbottema** [comment]
Found this library as well: https://github.com/usrflo/DKIM-for-JavaMail

**Comment by chaopeng** [comment]
Yes the one I use is fork on https://github.com/usrflo/DKIM-for-JavaMail . I use https://github.com/markenwerk/java-utils-mail-dkim is because DKIM-for-JavaMail is not in any public maven repo and no activity for many years.

**Comment by bbottema** [comment]
@chaopeng, I made some changes, but I don't know how to test this easily. Do you have a domain with selector and a sample key for me which I can use to test with?

**Comment by chaopeng** [comment]
@bbottema do you have domain? If yes you can use this script to generate a key. And selector will be 'test'

```
#!/usr/bin/env bash

openssl genrsa -out dkim.pem 1024
openssl pkcs8 -topk8 -nocrypt -in dkim.pem -outform der -out dkim.der
openssl rsa -in dkim.pem -pubout

echo "please set DNS TXT record 'test._domainkey.example.com' as 'v=DKIM1;k=rsa;p={publickey};s=email;t=s'"
```

If you dont I can setup my domain and pm you the key for temporary test.

**Comment by bbottema** [comment]
Hi @chaopeng, I committed my changes into the branch ["DKIM"](https://github.com/bbottema/simple-java-mail/tree/DKIM).

I already verified the initial stage of creating the dkim signer and parsing the private DER key works properly, but I have yet to test the actual sending and verification of the domain record.

If you checkout the branch DKIM, you can try it if you want. See [`MailerTest`](https://github.com/bbottema/simple-java-mail/blob/DKIM/src/test/java/org/codemonkey/simplejavamail/MailerTest.java) for the API in action, it's very simple. `Mailer.signWithDomainKey()` accepts a File, `Inputsource` or `byte[]`.

**Comment by bbottema** [comment]
I did some testing with my own domain and it's going quite well, actually, up to the point where the email is being signed (after successfully getting the domain record). When signing the email's content, I get an error:

> Caused by: net.markenwerk.utils.mail.dkim.DkimSigningException: Performing RSA cryptography failed.;
>   nested exception is:
> 	java.io.IOException: javax.crypto.IllegalBlockSizeException: Data must not be longer than 128 bytes
> 	at net.markenwerk.utils.mail.dkim.DomainKey.check(DomainKey.java:270)
> 	at net.markenwerk.utils.mail.dkim.DkimSigner.sign(DkimSigner.java:439)
> 	at net.markenwerk.utils.mail.dkim.DkimMessage.writeTo(DkimMessage.java:141)
> 	at com.sun.mail.smtp.SMTPTransport.sendMessage(SMTPTransport.java:1241)
> 	at org.codemonkey.simplejavamail.Mailer.sendMail(Mailer.java:247)
> 	... 28 more
> Caused by: java.io.IOException: javax.crypto.IllegalBlockSizeException: Data must not be longer than 128 bytes
> 	at net.markenwerk.utils.mail.dkim.DkimSigningException.<init>(DkimSigningException.java:77)
> 	... 33 more
> Caused by: javax.crypto.IllegalBlockSizeException: Data must not be longer than 128 bytes
> 	at com.sun.crypto.provider.RSACipher.doFinal(RSACipher.java:335)
> 	at com.sun.crypto.provider.RSACipher.engineDoFinal(RSACipher.java:380)
> 	at javax.crypto.Cipher.doFinal(Cipher.java:2121)
> 	at net.markenwerk.utils.mail.dkim.DomainKey.check(DomainKey.java:257)
> 	... 32 more

**Comment by bbottema** [comment]
Ok, cool. It works. This is what gmail received:

> X-Google-Original-From: lollypop <lol.pop@umbhali.com>
> DKIM-Signature: v=1; a=rsa-sha256; q=dns/txt; c=simple/relaxed; t=1462955593;
> 	s=select; d=umbhali.com; i=lol.pop@umbhali.com;
> 	h=Date:From:To:Message-ID:Subject:MIME-Version:Content-Type;
> 	l=2018; bh=NTpB2kzJrbcBgD27dzkFvL2ulM5fuqqy28iGT6qNOw4=;
> 	b=uxrUbv803HrG0WJDz6G4SvNAowobFwzVyVVr7HTNBWO5g94fiYvB2nyQl1Oto2Uf
> 	ZkYAvf2yajRTLOK0OMhR1KYOjA16SGKkSfWjck+2E2yzdn0kmTpDzv8Sg1A0A+adMNv
> 	0VHUDlfY3FtBEkV2Q1ye1xpXap52WxVZedl4CXug=

**Comment by bbottema** [comment]
Released in [v3.1.0](https://github.com/bbottema/simple-java-mail/releases/tag/v3.1.0).
